File: src/domain/repositories/ImageFolderRepository.ts
import type { ImageFolder } from '@/domain/entities/ImageFolder'
import type { GeoLocation } from '@/domain/value-objects/GeoLocation'

export interface CreateImageFolderInput {
  id?: string
  areaId: string
  areaName: string
  name: string
  description: string
  ownerId: string
  ownerName: string
  assignToAllTechnicians: boolean
  assignedTechnicianIds: string[]
  assignedTechnicianNames: string[]
  routeCode?: string
  location?: GeoLocation
}

export interface UpdateImageFolderInput {
  name?: string
  description?: string
  assignToAllTechnicians?: boolean
  assignedTechnicianIds?: string[]
  assignedTechnicianNames?: string[]
}

export interface ImageFolderRepository {
  getById(id: string): Promise<ImageFolder | null>
  listAll(): Promise<ImageFolder[]>
  listByOwner(ownerId: string): Promise<ImageFolder[]>
  /** Carpetas propias, asignadas o abiertas a todos los técnicos. */
  listAccessibleForUser(userId: string): Promise<ImageFolder[]>
  listByArea(areaId: string): Promise<ImageFolder[]>
  listWithoutArea(): Promise<ImageFolder[]>
  create(input: CreateImageFolderInput): Promise<ImageFolder>
  update(id: string, input: UpdateImageFolderInput): Promise<ImageFolder>
  assignArea(
    id: string,
    input: { areaId: string; areaName: string },
  ): Promise<void>
  incrementImageCount(folderId: string, delta: number): Promise<void>
  delete(id: string): Promise<void>
}

File: consorcio/lib/domain/repositories/image_folder_repository.dart
import '../entities/image_folder.dart';
import '../value_objects/geo_location.dart';

class CreateImageFolderInput {
  const CreateImageFolderInput({
    required this.areaId,
    required this.areaName,
    required this.name,
    required this.description,
    required this.ownerId,
    required this.ownerName,
    required this.assignToAllTechnicians,
    required this.assignedTechnicianIds,
    required this.assignedTechnicianNames,
    this.location,
  });

  final String areaId;
  final String areaName;
  final String name;
  final String description;
  final String ownerId;
  final String ownerName;
  final bool assignToAllTechnicians;
  final List<String> assignedTechnicianIds;
  final List<String> assignedTechnicianNames;
  final GeoLocation? location;
}

abstract class ImageFolderRepository {
  Future<ImageFolder?> getById(String id);
  Future<List<ImageFolder>> listByOwner(String ownerId);
  Future<List<ImageFolder>> listAccessibleForUser(String userId);
  Future<List<ImageFolder>> listByOwnerAndArea({
    required String ownerId,
    required String areaId,
  });
  Future<List<ImageFolder>> listAccessibleForUserAndArea({
    required String userId,
    required String areaId,
  });
  Future<ImageFolder> create(CreateImageFolderInput input);
  Future<ImageFolder> update({
    required String id,
    required String name,
    required String description,
    required bool assignToAllTechnicians,
    required List<String> assignedTechnicianIds,
    required List<String> assignedTechnicianNames,
  });
  Future<void> incrementImageCount(String folderId, int delta);
}
